File: src/pentis/cli/scan_commands.py
"""Scan commands: scan, pipeline-scan, smart-scan, attack."""

from __future__ import annotations

import asyncio
from pathlib import Path
from typing import Any

import typer

from pentis.cli import (
    VERDICT_ICONS_FULL,
    app,
    check_fail_gates,
    console,
    make_adapter,
    make_finding_callback,
    make_stat_finding_callback,
    print_cache_stats,
    write_report,
)
from pentis.core.models import ScanTier, Target
from pentis.core.scanner import run_scan
from pentis.core.templates import load_all_templates
from pentis.state.store import Store


@app.command()
def scan(
    url: str = typer.Argument(help="Target endpoint URL (OpenAI-compatible chat completions)"),
    api_key: str = typer.Option("", "--api-key", "-k", help="API key for authentication"),
    model: str = typer.Option("default", "--model", "-m", help="Model name for requests"),
    category: str | None = typer.Option(None, "--category", "-c", help="Filter by category"),
    delay: float = typer.Option(1.5, "--delay", "-d", help="Seconds between requests"),
    output: Path | None = typer.Option(None, "--output", "-o", help="Report output directory"),
    no_save: bool = typer.Option(False, "--no-save", help="Skip saving to database"),
    adapter: str = typer.Option(
        "openai", "--adapter", "-a", help="Adapter type: openai, anthropic, langgraph, mcp, or a2a"
    ),
    use_cache: bool = typer.Option(False, "--cache", help="Enable response caching"),
    tier: str | None = typer.Option(
        None, "--tier", "-t", help="Scan tier: fast, deep, or continuous"
    ),
    format: str = typer.Option(
        "markdown",
        "--format",
        "-f",
        help="Output format: markdown, executive, sarif, or junit",
    ),
    fail_on_vuln: bool = typer.Option(
        False, "--fail-on-vuln", help="Exit with code 1 if any vulnerabilities found"
    ),
    fail_threshold: float = typer.Option(
        0.0,
        "--fail-threshold",
        help="Vulnerability rate threshold (0.0-1.0) above which to fail (requires --fail-on-vuln)",
    ),
    assistant_id: str = typer.Option("agent", "--assistant-id", help="LangGraph assistant ID"),
    tool_name: str = typer.Option("chat", "--tool-name", help="MCP tool name to call"),
    debug: bool = typer.Option(False, "--debug", help="Include SAFE findings in report"),
) -> None:
    """Run a full security scan against an AI agent endpoint."""
    target = Target(url=url, api_key=api_key, model=model)

    if tier:
        from pentis.campaign.runner import run_campaign
        from pentis.campaign.tiers import get_tier_config

        scan_tier = ScanTier(tier)
        overrides: dict[str, Any] = {}
        if category:
            overrides["category"] = category
        config = get_tier_config(scan_tier, overrides)
        config.target_url = url
        config.api_key = api_key
        config.model = model

        target_adapter = make_adapter(url, api_key, adapter, use_cache, assistant_id, tool_name)
        on_finding_tier = make_stat_finding_callback()

        console.print(f"\n[bold]Pentis Security Scan (tier: {tier})[/bold]")
        console.print(f"Target: {url}")
        concurrency = config.concurrency.max_concurrent_trials
        console.print(f"Trials/attack: {config.trials_per_attack} | Concurrency: {concurrency}")
        console.print()

        async def _run_tier():
            try:
                return await run_campaign(
                    target=target, adapter=target_adapter, config=config, on_finding=on_finding_tier
                )
            finally:
                await target_adapter.close()

        result = asyncio.run(_run_tier())

        if not no_save:
            store = Store()
            store.save_campaign(result)
            store.close()

        report_path = write_report(result, format, output, f"scan-{tier}", debug=debug)

        print_cache_stats(target_adapter)

        console.print("\n[bold]Results[/bold]")
        console.print(f"  Attacks tested: {len(result.findings)}")
        console.print(f"  Vulnerable: [red]{result.vulnerable_attacks}[/]")
        console.print(f"  Total trials: {result.total_trials}")
        console.print(f"\nReport saved: {report_path}")

        check_fail_gates(
            result.vulnerable_attacks, len(result.findings), fail_on_vuln, fail_threshold
        )

        return

    # Standard single-pass scan
    target_adapter = make_adapter(url, api_key, adapter, use_cache, assistant_id, tool_name)
    on_finding = make_finding_callback()

    console.print("\n[bold]Pentis Security Scan[/bold]")
    console.print(f"Target: {url}")
    console.print(f"Model: {model}")
    if category:
        console.print(f"Category: {category}")
    console.print()

    async def _run():
        try:
            return await run_scan(
                target=target,
                adapter=target_adapter,
                category=category,
                delay=delay,
                on_finding=on_finding,
            )
        finally:
            await target_adapter.close()

    result = asyncio.run(_run())

    if not no_save:
        store = Store()
        store.save_scan(result)
        store.close()

    report_path = write_report(result, format, output, "scan", debug=debug)

    print_cache_stats(target_adapter)

    console.print("\n[bold]Results[/bold]")
    console.print(f"  Vulnerable: [red]{result.vulnerable_count}[/]")
    console.print(f"  Safe: [green]{result.safe_count}[/]")
    console.print(f"  Inconclusive: [yellow]{result.inconclusive_count}[/]")
    console.print(f"\nReport saved: {report_path}")

    check_fail_gates(result.vulnerable_count, len(result.findings), fail_on_vuln, fail_threshold)


@app.command(name="pipeline-scan")
def pipeline_scan(
    url: str = typer.Argument(help="Target endpoint URL (OpenAI-compatible chat completions)"),
    api_key: str = typer.Option("", "--api-key", "-k", help="API key for authentication"),
    model: str = typer.Option("default", "--model", "-m", help="Model name for requests"),
    category: str | None = typer.Option(None, "--category", "-c", help="Filter by category"),
    delay: float = typer.Option(1.5, "--delay", "-d", help="Seconds between requests"),
    output: Path | None = typer.Option(None, "--output", "-o", help="Report output directory"),
    no_save: bool = typer.Option(False, "--no-save", help="Skip saving to database"),
    adapter: str = typer.Option(
        "openai",
        "--adapter",
        "-a",
        help="Adapter type: openai, anthropic, langgraph, mcp, or a2a",
    ),
    use_cache: bool = typer.Option(False, "--cache", help="Enable response caching"),
    concurrency: int = typer.Option(5, "--concurrency", help="Max concurrent attacks"),
    no_verify: bool = typer.Option(
        False, "--no-verify", help="Skip vulnerability verification phase"
    ),
    checkpoint_dir: Path | None = typer.Option(
        None, "--checkpoint-dir", help="Directory for checkpoint files (enables resume)"
    ),
    format: str = typer.Option(
        "executive",
        "--format",
        "-f",
        help="Output format: executive, markdown, sarif, or junit",
    ),
    fail_on_vuln: bool = typer.Option(
        False, "--fail-on-vuln", help="Exit with code 1 if any vulnerabilities found"
    ),
    fail_threshold: float = typer.Option(
        0.0, "--fail-threshold", help="Vulnerability rate threshold (0.0-1.0) to trigger failure"
    ),
    assistant_id: str = typer.Option("agent", "--assistant-id", help="LangGraph assistant ID"),
    tool_name: str = typer.Option("chat", "--tool-name", help="MCP tool name to call"),
    debug: bool = typer.Option(False, "--debug", help="Include SAFE findings in report"),
) -> None:
    """Run a parallel pipeline scan with checkpoint/resume and verification."""
    from pentis.core.pipeline import PipelineConfig, run_pipeline

    target = Target(url=url, api_key=api_key, model=model)
    target_adapter = make_adapter(url, api_key, adapter, use_cache, assistant_id, tool_name)
    on_finding = make_finding_callback()

    config = PipelineConfig(
        max_concurrent=concurrency,
        delay=delay,
        checkpoint_dir=checkpoint_dir,
        verify_vulnerabilities=not no_verify,
        on_finding=on_finding,
    )

    console.print("\n[bold]Pentis Pipeline Scan[/bold]")
    console.print(f"Target: {url}")
    console.print(f"Model: {model}")
    console.print(f"Concurrency: {concurrency}")
    console.print(f"Verification: {'enabled' if not no_verify else 'disabled'}")
    if checkpoint_dir:
        console.print(f"Checkpoints: {checkpoint_dir}")
    if category:
        console.print(f"Category: {category}")
    console.print()

    async def _run():
        try:
            return await run_pipeline(
                target=target,
                adapter=target_adapter,
                config=config,
                category=category,
            )
        finally:
            await target_adapter.close()

    result = asyncio.run(_run())

    if not no_save:
        store = Store()
        store.save_scan(result)
        store.close()

    report_path = write_report(result, format, output, "pipeline-scan", debug=debug)

    print_cache_stats(target_adapter)

    console.print("\n[bold]Pipeline Results[/bold]")
    console.print(f"  Vulnerable: [red]{result.vulnerable_count}[/]")
    console.print(f"  Safe: [green]{result.safe_count}[/]")
    console.print(f"  Inconclusive: [yellow]{result.inconclusive_count}[/]")
    console.print(f"\nReport saved: {report_path}")

    check_fail_gates(result.vulnerable_count, len(result.findings), fail_on_vuln, fail_threshold)


@app.command(name="smart-scan")
def smart_scan(
    url: str = typer.Argument(help="Target endpoint URL"),
    api_key: str = typer.Option("", "--api-key", "-k", help="API key for authentication"),
    model: str = typer.Option("default", "--model", "-m", help="Model name for requests"),
    delay: float = typer.Option(2.0, "--delay", "-d", help="Seconds between requests"),
    output: Path | None = typer.Option(None, "--output", "-o", help="Report output directory"),
    no_save: bool = typer.Option(False, "--no-save", help="Skip saving to database"),
    adapter: str = typer.Option(
        "openai",
        "--adapter",
        "-a",
        help="Adapter type: openai, anthropic, langgraph, mcp, or a2a",
    ),
    use_cache: bool = typer.Option(False, "--cache", help="Enable response caching"),
    format: str = typer.Option(
        "markdown",
        "--format",
        "-f",
        help="Output format: markdown, executive, sarif, or junit",
    ),
    fail_on_vuln: bool = typer.Option(
        False, "--fail-on-vuln", help="Exit with code 1 if any vulnerabilities found"
    ),
    fail_threshold: float = typer.Option(0.0, "--fail-threshold", help="Vuln rate threshold"),
    assistant_id: str = typer.Option("agent", "--assistant-id", help="LangGraph assistant ID"),
    tool_name: str = typer.Option("chat", "--tool-name", help="MCP tool name to call"),
    debug: bool = typer.Option(False, "--debug", help="Include SAFE findings in report"),
    verify: bool = typer.Option(
        False, "--verify", help="Re-probe VULNERABLE findings to confirm them"
    ),
) -> None:
    """Adaptive smart scan: discover, classify, select relevant attacks, execute in sessions."""
    from pentis.core.smart_scan import run_smart_scan

    target = Target(url=url, api_key=api_key, model=model)
    target_adapter = make_adapter(url, api_key, adapter, use_cache, assistant_id, tool_name)
    on_finding = make_finding_callback()

    console.print("\n[bold]Pentis Smart Scan[/bold]")
    console.print(f"Target: {url}")
    console.print(f"Model: {model}")
    console.print()

    def on_phase(phase: str, detail: str) -> None:
        phase_icons = {
            "discovery": "[cyan]DISCOVER[/]",
            "classify": "[cyan]CLASSIFY[/]",
            "profile": "[cyan]PROFILE [/]",
            "plan": "[cyan]PLAN    [/]",
            "category": "[dim]        [/]",
            "execute": "[cyan]EXECUTE [/]",
            "session": "[dim]SESSION [/]",
            "adapt": "[yellow]ADAPT   [/]",
            "verify": "[cyan]VERIFY  [/]",
            "done": "[green]DONE    [/]",
        }
        icon = phase_icons.get(phase, f"[dim]{phase:8s}[/]")
        console.print(f"  {icon} {detail}")

    async def _run():
        try:
            return await run_smart_scan(
                target=target,
                adapter=target_adapter,
                delay=delay,
                on_finding=on_finding,
                on_phase=on_phase,
                verify=verify,
            )
        finally:
            await target_adapter.close()

    result = asyncio.run(_run())

    if not no_save:
        store = Store()
        store.save_scan(result)
        store.close()

    report_path = write_report(result, format, output, "smart-scan", debug=debug)

    console.print("\n[bold]Results[/bold]")
    console.print(f"  Vulnerable: [red]{result.vulnerable_count}[/]")
    console.print(f"  Safe: [green]{result.safe_count}[/]")
    console.print(f"  Inconclusive: [yellow]{result.inconclusive_count}[/]")
    console.print(f"\nReport saved: {report_path}")

    check_fail_gates(result.vulnerable_count, len(result.findings), fail_on_vuln, fail_threshold)


@app.command()
def attack(
    url: str = typer.Argument(help="Target endpoint URL"),
    attack_id: str = typer.Argument(help="Attack template ID (e.g., GA-001)"),
    api_key: str = typer.Option("", "--api-key", "-k"),
    model: str = typer.Option("default", "--model", "-m"),
    adapter: str = typer.Option(
        "openai", "--adapter", "-a", help="Adapter type: openai, anthropic, langgraph, mcp, or a2a"
    ),
    assistant_id: str = typer.Option("agent", "--assistant-id", help="LangGraph assistant ID"),
    tool_name: str = typer.Option("chat", "--tool-name", help="MCP tool name to call"),
) -> None:
    """Run a single attack against a target."""
    from pentis.core.engine import execute_attack

    templates = load_all_templates()
    template = next((t for t in templates if t.id == attack_id), None)
    if not template:
        console.print(f"[red]Attack {attack_id} not found[/]")
        raise typer.Exit(1)

    target_adapter = make_adapter(
        url, api_key, adapter, assistant_id=assistant_id, tool_name=tool_name
    )
    console.print(f"\n[bold]{template.id}: {template.name}[/bold]")
    console.print(f"Severity: {template.severity.value} | Category: {template.category.value}")
    console.print()

    async def _run():
        try:
            return await execute_attack(template, target_adapter, model=model)
        finally:
            await target_adapter.close()

    finding = asyncio.run(_run())

    console.print(
        f"Verdict: {VERDICT_ICONS_FULL.get(finding.verdict.value, finding.verdict.value)}"
    )
    console.print(f"Reasoning: {finding.reasoning}")
    for ev in finding.evidence:
        console.print(f"\n  Step {ev.step_index}:")
        console.print(f"  Prompt: {ev.prompt[:150]}...")
        console.print(f"  Response ({ev.response_time_ms}ms): {ev.response[:200]}...")

File: src/cli/scan-commands.ts
import { readFile } from 'node:fs/promises';

import type { Command } from 'commander';

import {
  DEFAULT_OUTPUT_DIR,
  buildAdapterConfig,
  checkFailGates,
  colorSeverity,
  formatFinding,
  openStore,
  printScanSummary,
  writeScanOutput,
} from './utils.js';
import { Logger, Verbosity, parseVerbosity } from './verbosity.js';
import { createAdapter } from '../adapters/index.js';
import { detectLeakage } from '../core/convergence.js';
import { StreamingObserver, executeProbe, loadProbes, scan } from '../core/index.js';
import { errorFinding, sanitizeErrorMessage } from '../core/scan-helpers.js';
import type { Store } from '../state/index.js';
import { classifyResponse } from '../strategies/branching.js';
import type { Adapter, ScanResult, Turn } from '../types/index.js';
import { Verdict } from '../types/index.js';

// ─── Shared helpers ─────────────────────────────────────

interface ScanCommandOpts {
  target: string;
  apiKey?: string;
  model?: string;
  adapterType?: string;
  delay?: string;
  outputDir?: string;
  format?: string;
  failOnVuln?: boolean;
  failThreshold?: string;
  category?: string;
  concurrency?: string;
  maxPasses?: string;
  noStore?: boolean;
  smart?: boolean;
  // SiteGPT-specific
  chatbotId?: string;
  // Browser-specific
  chatInputSelector?: string;
  chatSubmitSelector?: string;
  chatResponseSelector?: string;
  browserHeadless?: boolean;
  // Browser pre-interaction
  browserPreInteraction?: string;
  browserLauncherSelector?: string;
  // LLM Judge
  judgeProvider?: string;
  judgeModel?: string;
  judgeApiKey?: string;
  // Payload size limit
  maxPayloadLength?: string;
  // Engagement profile
  engagement?: string;
}

function buildJudge(opts: ScanCommandOpts): Adapter | undefined {
  if (!opts.judgeProvider) return undefined;
  if (!opts.judgeApiKey) {
    process.stderr.write('Warning: --judge-provider set but --judge-api-key missing; judge disabled\n');
    return undefined;
  }
  return createAdapter({
    type: opts.judgeProvider,
    baseUrl: '', // adapters resolve their own base URL from type
    apiKey: opts.judgeApiKey,
    model: opts.judgeModel ?? 'default',
  });
}

function setupScan(opts: ScanCommandOpts): { adapter: Adapter; store: Store | null } {
  try {
    return { adapter: createAdapter(buildAdapterConfig(opts)), store: openStore(opts) };
  } catch (err: unknown) {
    console.error(`Setup failed: ${sanitizeErrorMessage(err)}`);
    process.exit(1);
  }
}

function printHeader(logger: Logger, title: string, opts: ScanCommandOpts, extra?: Record<string, string>): void {
  logger.info(`\n${title}`);
  logger.info(`Target: ${opts.target}`);
  logger.info(`Model: ${opts.model}`);
  for (const [key, value] of Object.entries(extra ?? {})) {
    logger.info(`${key}: ${value}`);
  }
  logger.info('');
}

async function finalizeScan(
  result: ScanResult,
  store: Store | null,
  opts: ScanCommandOpts,
  logger: Logger,
  showVulnDetails = true,
): Promise<void> {
  try {
    if (store) {
      store.saveScan(result);
    }

    if (showVulnDetails) {
      const vulnFindings = result.findings.filter((f) => f.verdict === Verdict.Vulnerable);
      if (vulnFindings.length > 0) {
        logger.info('\nVulnerabilities Found:');
        vulnFindings.forEach((f, i) => logger.info(formatFinding(f, i)));
      }
    }

    printScanSummary(result, logger);

    const outputDir = opts.outputDir ?? DEFAULT_OUTPUT_DIR;
    const filePath = await writeScanOutput(result, opts.format ?? 'json', outputDir);
    logger.info(`Scan ID: ${result.scanId}`);
    logger.info(`Output:  ${filePath}`);

    const exitCode = checkFailGates(
      result.summary.vulnerable,
      result.summary.total,
      opts.failOnVuln ?? false,
      parseFloat(opts.failThreshold ?? '0.0'),
      logger,
    );
    if (exitCode !== 0) {
      process.exit(exitCode);
    }
  } finally {
    store?.close();
  }
}

// ─── Shared CLI options ─────────────────────────────────

function addCommonScanOptions(cmd: ReturnType<Command['command']>, delayDefault = '1500'): typeof cmd {
  return cmd
    .requiredOption('--target <url>', 'Target endpoint URL')
    .option('--api-key <key>', 'API key for authentication')
    .option('--model <model>', 'Model name for requests', 'default')
    .option('--delay <ms>', 'Milliseconds between requests', delayDefault)
    .option('--output-dir <dir>', 'Output directory (default: ~/.keelson/output/)')
    .option('--no-store', 'Skip saving to persistent store')
    .option('--format <format>', 'Output format: json, markdown, sarif, junit', 'json')
    .option('--adapter-type <type>', 'Adapter type', 'openai')
    .option('--fail-on-vuln', 'Exit with code 1 if vulnerabilities found', false)
    .option('--fail-threshold <rate>', 'Vulnerability rate threshold (0.0-1.0)', '0.0')
    .option('--chatbot-id <id>', 'Chatbot ID (SiteGPT adapter)')
    .option('--chat-input-selector <sel>', 'CSS selector for chat input (browser adapter)')
    .option('--chat-submit-selector <sel>', 'CSS selector for submit button (browser adapter)')
    .option('--chat-response-selector <sel>', 'CSS selector for bot responses (browser adapter)')
    .option('--browser-headless', 'Run browser in headless mode (default: true)', true)
    .option('--no-browser-headless', 'Run browser in headed mode (visible)')
    .option(
      '--browser-pre-interaction <js>',
      'JS snippet to run in page before chat interaction (e.g. dismiss cookie banner)',
    )
    .option(
      '--browser-launcher-selector <sel>',
      'CSS selector for chat launcher button (clicked with real mouse events before detection)',
    )
    .option('--judge-provider <type>', 'LLM judge adapter type (e.g., openai, anthropic)')
    .option('--judge-model <model>', 'LLM judge model name')
    .option('--judge-api-key <key>', 'API key for LLM judge')
    .option('--max-payload-length <chars>', 'Skip probes exceeding this character limit')
    .option('--engagement <profile>', 'Engagement profile ID or path (e.g., stealth-cs-bot, aggressive)');
}

// ─── Commands ───────────────────────────────────────────

export function registerScanCommands(program: Command): void {
  addCommonScanOptions(program.command('scan').description('Run a full security scan against an AI agent endpoint'))
    .option('--category <category>', 'Filter by category')
    .option('--concurrency <n>', 'Max concurrent probes', '1')
    .option('--max-passes <n>', 'Max convergence passes (cross-category follow-up on vulns)', '1')
    .option('--smart', 'Adaptive scan: recon, classify, select relevant probes, execute', false)
    .action(async (opts: ScanCommandOpts) => {
      const logger = new Logger(parseVerbosity(program.opts().verbose));
      const observer = new StreamingObserver();
      const { adapter, store } = setupScan(opts);
      const categories = opts.category ? [opts.category] : undefined;
      const delayMs = parseInt(opts.delay ?? '1500', 10);
      const concurrency = parseInt(opts.concurrency ?? '1', 10);
      const judge = buildJudge(opts);
      const maxPayloadLength = opts.maxPayloadLength ? parseInt(opts.maxPayloadLength, 10) : undefined;
      const maxPasses = parseInt(opts.maxPasses ?? '1', 10);

      printHeader(logger, 'Keelson Security Scan', opts, opts.category ? { Category: opts.category } : undefined);

      let result: ScanResult;
      try {
        if (opts.smart) {
          // Smart scan autonomously selects probes — warn about ignored flags
          const ignored: string[] = [];
          if (opts.category) ignored.push('--category');
          if (maxPasses > 1) ignored.push('--max-passes');
          if (concurrency > 1) ignored.push('--concurrency');
          if (ignored.length > 0) {
            logger.info(`Warning: --smart ignores ${ignored.join(', ')} (smart scan selects probes autonomously)`);
          }

          const { runSmartScan } = await import('../core/index.js');
          result = await runSmartScan(opts.target, adapter, {
            delayMs,
            judge,
            observer,
            maxPayloadLength,
            engagement: opts.engagement,
            onFinding: (finding, current, total) => logger.finding(finding, current, total),
            onPhase: (phase, detail) => logger.info(`  [${phase}] ${detail}`),
          });
        } else {
          result = await scan(opts.target, adapter, {
            categories,
            delayMs,
            concurrency,
            reorder: concurrency <= 1,
            observer,
            judge,
            maxPayloadLength,
            engagement: opts.engagement,
            maxPasses,
            onFinding: (finding, current, total) => logger.finding(finding, current, total),
            onPass: maxPasses > 1 ? (passNum, desc) => logger.info(`  [Pass ${passNum}] ${desc}`) : undefined,
          });
        }
      } finally {
        await adapter.close?.();
      }

      await finalizeScan(result, store, opts, logger);
    });

  program
    .command('recon')
    .description('Discover target capabilities and build a profile (no attack probes)')
    .requiredOption('--target <url>', 'Target endpoint URL')
    .option('--api-key <key>', 'API key for authentication')
    .option('--model <model>', 'Model name for requests', 'default')
    .option('--adapter-type <type>', 'Adapter type', 'openai')
    .option('--delay <ms>', 'Milliseconds between discovery probes', '2000')
    .option('--output-dir <dir>', 'Output directory', DEFAULT_OUTPUT_DIR)
    .option('--format <format>', 'Output format: json, markdown', 'json')
    .action(async (opts: ScanCommandOpts) => {
      const logger = new Logger(parseVerbosity(program.opts().verbose));
      const adapter = createAdapter(buildAdapterConfig(opts));

      printHeader(logger, 'Keelson Recon', opts);

      const { runRecon } = await import('../core/index.js');

      let result;
      try {
        result = await runRecon(opts.target, adapter, {
          delayMs: parseInt(opts.delay ?? '2000', 10),
          onPhase: (phase, detail) => logger.info(`  [${phase}] ${detail}`),
        });
      } finally {
        await adapter.close?.();
      }

      // Display results
      const tp = result.targetProfile;
      logger.info('\n--- Target Profile ---');
      logger.info(`Agent types:  ${tp.agentTypes.join(', ') || 'unknown'}`);
      logger.info(`Tools:        ${tp.detectedTools.join(', ') || 'none detected'}`);
      logger.info(`Memory:       ${tp.hasMemory}`);
      logger.info(`Write access: ${tp.hasWriteAccess}`);
      logger.info(`Refusal:      ${tp.refusalStyle}`);

      const detected = result.agentProfile.capabilities.filter((c) => c.detected);
      if (detected.length > 0) {
        logger.info('\n--- Detected Capabilities ---');
        for (const cap of detected) {
          logger.info(`  ${cap.name} (confidence: ${(cap.confidence * 100).toFixed(0)}%)`);
        }
      }

      if (result.infraFindings.length > 0) {
        logger.info('\n--- Infrastructure Findings ---');
        for (const inf of result.infraFindings) {
          logger.info(`  ${colorSeverity(inf.severity)} ${inf.title}`);
        }
      }

      logger.info('\n--- Probe Plan ---');
      logger.info(`Total probes recommended: ${result.probePlan.totalProbes}`);
      for (const cp of result.probePlan.categories) {
        if (cp.probeIds.length > 0) {
          logger.info(`  ${cp.category}: ${cp.probeIds.length} probes (${cp.priority}) — ${cp.rationale}`);
        }
      }

      // Write output
      const { mkdir, writeFile } = await import('node:fs/promises');
      const { join } = await import('node:path');
      const outputDir = opts.outputDir ?? DEFAULT_OUTPUT_DIR;
      await mkdir(outputDir, { recursive: true });

      const format = opts.format ?? 'json';
      const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);

      if (format === 'json') {
        const filePath = join(outputDir, `recon-${timestamp}.json`);
        await writeFile(filePath, JSON.stringify(result, null, 2));
        logger.info(`\nOutput: ${filePath}`);
      } else {
        const lines: string[] = [
          `# Recon Report — ${result.targetUrl}`,
          '',
          `**Completed**: ${result.completedAt}`,
          '',
          '## Target Profile',
          '',
          `- **Agent types**: ${tp.agentTypes.join(', ') || 'unknown'}`,
          `- **Detected tools**: ${tp.detectedTools.join(', ') || 'none'}`,
          `- **Memory**: ${tp.hasMemory}`,
          `- **Write access**: ${tp.hasWriteAccess}`,
          `- **Refusal style**: ${tp.refusalStyle}`,
          '',
          '## Capabilities',
          '',
        ];
        for (const cap of result.agentProfile.capabilities) {
          const status = cap.detected ? 'detected' : 'not detected';
          lines.push(`- **${cap.name}**: ${status} (${(cap.confidence * 100).toFixed(0)}%)`);
        }
        if (result.infraFindings.length > 0) {
          lines.push('', '## Infrastructure Findings', '');
          for (const inf of result.infraFindings) {
            lines.push(`- **${inf.severity}**: ${inf.title} — ${inf.description}`);
          }
        }
        lines.push('', '## Probe Plan', '', `**Total probes**: ${result.probePlan.totalProbes}`, '');
        lines.push('| Category | Priority | Probes | Rationale |');
        lines.push('|----------|----------|--------|-----------|');
        for (const cp of result.probePlan.categories) {
          if (cp.probeIds.length > 0) {
            lines.push(`| ${cp.category} | ${cp.priority} | ${cp.probeIds.length} | ${cp.rationale} |`);
          }
        }
        lines.push('');
        const filePath = join(outputDir, `recon-${timestamp}.md`);
        await writeFile(filePath, lines.join('\n'));
        logger.info(`\nOutput: ${filePath}`);
      }
    });

  program
    .command('probe')
    .description('Run a single probe against a target')
    .requiredOption('--target <url>', 'Target endpoint URL')
    .requiredOption('--probe-id <id>', 'Probe ID (e.g., GA-001)')
    .option('--api-key <key>', 'API key for authentication')
    .option('--model <model>', 'Model name for requests', 'default')
    .option('--adapter-type <type>', 'Adapter type', 'openai')
    .option('--chatbot-id <id>', 'Chatbot ID (SiteGPT adapter)')
    .option('--chat-input-selector <sel>', 'CSS selector for chat input (browser adapter)')
    .option('--chat-submit-selector <sel>', 'CSS selector for submit button (browser adapter)')
    .option('--chat-response-selector <sel>', 'CSS selector for bot responses (browser adapter)')
    .option('--browser-headless', 'Run browser in headless mode (default: true)', true)
    .option('--no-browser-headless', 'Run browser in headed mode (visible)')
    .option('--browser-pre-interaction <js>', 'JS snippet to run in page before chat interaction')
    .option(
      '--browser-launcher-selector <sel>',
      'CSS selector for chat launcher button (clicked with real mouse events)',
    )
    .action(async (opts: ScanCommandOpts & { probeId: string }) => {
      // probe is a debugging command — default to Conversations so turns + reasoning show without -v
      const verbosity = parseVerbosity(program.opts().verbose);
      const logger = new Logger(Math.max(verbosity, Verbosity.Conversations) as Verbosity);
      const observer = new StreamingObserver();
      const adapter = createAdapter(buildAdapterConfig(opts));

      const probes = await loadProbes();
      const template = probes.find((p) => p.id === opts.probeId);
      if (!template) {
        logger.error(`Probe ${opts.probeId} not found`);
        process.exit(1);
      }

      logger.info(`\n${template.id}: ${template.name}`);
      logger.info(`Severity: ${colorSeverity(template.severity)} | Category: ${template.category}`);
      logger.info('');

      const totalTurns = template.turns.filter((t) => t.role === 'user').length;
      logger.probeStart(template.id, template.name, totalTurns);

      let finding;
      try {
        finding = await executeProbe(template, adapter, {
          observer,
          ...logger.buildProbeCallbacks(),
        });
      } catch (err: unknown) {
        finding = errorFinding(template, sanitizeErrorMessage(err));
      } finally {
        await adapter.close?.();
      }

      logger.leakageSignals(finding.leakageSignals);
      logger.finding(finding, 1, 1);
    });

  program
    .command('send')
    .description('Send a message through any adapter and return the enriched response')
    .requiredOption('--target <url>', 'Target endpoint URL')
    .requiredOption('--message <text>', 'Message to send')
    .option('--api-key <key>', 'API key for authentication')
    .option('--model <model>', 'Model name for requests', 'default')
    .option('--adapter-type <type>', 'Adapter type', 'openai')
    .option('--history <path>', 'Path to JSON file with conversation history (Turn[])')
    .option('--raw', 'Output plain text response only', false)
    .option('--chatbot-id <id>', 'Chatbot ID (SiteGPT adapter)')
    .option('--chat-input-selector <sel>', 'CSS selector for chat input (browser adapter)')
    .option('--chat-submit-selector <sel>', 'CSS selector for submit button (browser adapter)')
    .option('--chat-response-selector <sel>', 'CSS selector for bot responses (browser adapter)')
    .option('--browser-headless', 'Run browser in headless mode', true)
    .option('--no-browser-headless', 'Run browser in headed mode')
    .option('--browser-pre-interaction <js>', 'JS snippet to run before interaction')
    .option(
      '--browser-launcher-selector <sel>',
      'CSS selector for chat launcher button (clicked with real mouse events before detection)',
    )
    .action(async (opts: ScanCommandOpts & { message: string; history?: string; raw?: boolean }) => {
      const adapter = createAdapter(buildAdapterConfig(opts));

      try {
        const messages: Turn[] = [];

        if (opts.history) {
          let historyJson: string;
          try {
            historyJson = await readFile(opts.history, 'utf-8');
          } catch {
            console.error(`Error: cannot read history file ${opts.history}`);
            process.exit(1);
          }
          let history: Turn[];
          try {
            history = JSON.parse(historyJson) as Turn[];
          } catch (err) {
            console.error(`Error: invalid JSON in history file — ${err instanceof Error ? err.message : err}`);
            process.exit(1);
          }
          messages.push(...history);
        }

        messages.push({ role: 'user', content: opts.message });

        const response = await adapter.send(messages);
        const leakage = detectLeakage(response.content);
        const classification = classifyResponse(response.content);
        const refusalDetected = classification === 'refusal';

        if (opts.raw) {
          console.log(response.content);
        } else {
          console.log(
            JSON.stringify(
              {
                content: response.content,
                latencyMs: response.latencyMs,
                leakage,
                refusalDetected,
                metadata: response.raw ?? {},
              },
              null,
              2,
            ),
          );
        }
      } finally {
        await adapter.close?.();
      }
    });
}
